Gems/Atom/Tools/ShaderManagementConsole/Code/Source/Window/ShaderManagementConsoleTableView.cpp

```diff
@@ -49,56 +50,129 @@ namespace ShaderManagementConsole
+        // Only clear the table if the number of columns or rows have changed
+        if (m_model->rowCount() != shaderVariantCount || m_model->columnCount() != shaderOptionCount)
         {
-            AZ::RPI::ShaderOptionDescriptor shaderOptionDesc;
-            ShaderManagementConsoleDocumentRequestBus::EventResult(
-                shaderOptionDesc, m_documentId, &ShaderManagementConsoleDocumentRequestBus::Events::GetShaderOptionDescriptor, optionIndex);
-            optionNames.insert(shaderOptionDesc.GetName().GetCStr());
+            m_model->clear();
+            m_model->setRowCount(static_cast<int>(shaderVariantCount));
+            m_model->setColumnCount(static_cast<int>(shaderOptionCount));
         }
 
-        size_t shaderVariantCount = 0;
-        ShaderManagementConsoleDocumentRequestBus::EventResult(
-            shaderVariantCount, m_documentId, &ShaderManagementConsoleDocumentRequestBus::Events::GetShaderVariantCount);
+        // Get a list of all of the shader option descriptors from the shader asset that will be used for the columns in the table
+        AZStd::vector<AZ::RPI::ShaderOptionDescriptor> shaderOptionDescriptors;
+        for (size_t shaderOptionIndex = 0; shaderOptionIndex < shaderOptionCount; ++shaderOptionIndex)
+        {
+            AZ::RPI::ShaderOptionDescriptor shaderOptionDescriptor;
+            ShaderManagementConsoleDocumentRequestBus::EventResult(
+                shaderOptionDescriptor, m_documentId, &ShaderManagementConsoleDocumentRequestBus::Events::GetShaderOptionDescriptor,
+                shaderOptionIndex);
+            shaderOptionDescriptors.push_back(shaderOptionDescriptor);
+        }
 
-        m_model->clear();
-        m_model->setRowCount(static_cast<int>(shaderVariantCount));
-        m_model->setColumnCount(static_cast<int>(optionNames.size()));
+        // Sort all of the descriptors by name so that the columns are arranged in alphabetical order
+        AZStd::sort(shaderOptionDescriptors.begin(), shaderOptionDescriptors.end(), [](const auto& a, const auto& b) {
+            return a.GetName().GetStringView() < b.GetName().GetStringView();
+        });
 
-        int nameIndex = 0;
-        for (const auto& optionName : optionNames)
+        // Do a first pass iteration over all of the variants to set the numeric headers for each variant
+        size_t shaderVariantIndex = 0;
+        for (const auto& shaderVariant : shaderVariantListSourceData.m_shaderVariants)
         {
-            m_model->setHeaderData(nameIndex++, Qt::Horizontal, optionName.c_str());
+            m_model->setHeaderData(
+                static_cast<int>(shaderVariantIndex), Qt::Vertical, QString::number(static_cast<int>(shaderVariant.m_stableId)));
+            ++shaderVariantIndex;
         }
 
-        for (int variantIndex = 0; variantIndex < shaderVariantCount; ++variantIndex)
+        for (size_t shaderOptionIndex = 0; shaderOptionIndex < shaderOptionCount; ++shaderOptionIndex)
```

Comment:
I can imagine. I set it up this way so that I didn't have to fill in the header data for rows and columns in a separate loop. Before I disabled refreshing the treeview and sending the notifications to the document, which updated the inspector, you could clearly see that it took about a second to fill in each cell for something like standard PBR forward pass. It's pretty fast now but I could see it taking up time with large datasets. There might be a way to parallelize it or we could build a custom item model which would save us doing any of the iteration ourselves. That would also allow finer grained handling of individual cell changes instead of rebuilding the entire variant list with every change. It's still responsive as it is.

Reply:
Hm, yeah I could see the value in flipping columns and rows.